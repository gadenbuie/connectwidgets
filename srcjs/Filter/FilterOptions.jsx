import PropTypes from 'prop-types';
import { FilterContext } from './stateUtils';
import './filterOptions.scss';

const checkboxClass = {
  base: 'rscfilter-option__checkbox',
  checked: 'rscfilter-option__checkbox rscfilter-option__checkbox--checked',
};

function FilterOptions({ type }) {
  const context = React.useContext(FilterContext);
  const options = context.stateOptionsFor(type);

  const handleCheckbox = (ev, index) => {
    // Space key press is the standard to toggle checkboxes
    if(ev.key && ev.code !== 'Space') {
      return;
    }
    context.updateSelection({
      ...options[index],
      isSelected: !options[index].isSelected,
    });
  };

  const handleEsc = ev => {
    if (ev.key === "Escape") {
      ev.stopPropagation();
      document.activeElement.blur();
    }
  }

  return (
    <div className="rscfilter-field__options bg-white border">
      {
        options.map((option, index) => {
          const chClass = option.isSelected ? checkboxClass.checked : checkboxClass.base;
          return (
            <div
              key={option.name}
              className="rscfilter-option"
              onClick={ev => handleCheckbox(ev, index)}
              onKeyUp={handleEsc}
              onKeyPress={ev => handleCheckbox(ev, index)}
              role="checkbox"
              aria-checked={option.isSelected}
              tabIndex="0"
            >
              <div className={chClass} />
              <span className="rscfilter-option__label">
                {option.label}
              </span>
            </div>
          );
        })
      }
    </div>
  );
}

FilterOptions.propTypes = {
  type: PropTypes.string,
}

export default FilterOptions;
