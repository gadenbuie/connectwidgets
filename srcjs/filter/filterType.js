/**
 * Filter types
 */
 export const FilterType = {
  owner: 'owner',
  type: 'type',
  tag: 'tag',
};
