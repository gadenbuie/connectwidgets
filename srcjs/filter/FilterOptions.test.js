import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import FilterOptions from './FilterOptions';
import { FilterType } from './filterType';

const data = [
  {
    name: 'API',
    label: 'API',
    isSelected: false,
  },
  {
    name: 'Application',
    label: 'Application',
    isSelected: false,
  },
  {
    name: 'Document',
    label: 'Document',
    isSelected: false,
  },
  {
    name: 'Model',
    label: 'Model',
    isSelected: false,
  },
  {
    name: 'Pin',
    label: 'Pin',
    isSelected: false,
  },
];
const fakeContext = {
  options: { [FilterType.type]: data },
  stateOptionsFor: jest.fn().mockImplementation(() => data),
  updateSelection: jest.fn(),
};
jest.spyOn(React, 'useContext').mockReturnValue(fakeContext);

const mkWrapper = () => shallow(<FilterOptions
  type={FilterType.type}
/>);

describe('<FilterOptions />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('renders properly', () => {
    const wrapper = mkWrapper();
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  
  it('emits changes to context/state', () => {
    const wrapper = mkWrapper();
    const options = wrapper.find('.rscfilter-option');
    
    expect(fakeContext.updateSelection).not.toHaveBeenCalled();
    options.at(1).simulate('click', {});
    expect(fakeContext.updateSelection).toHaveBeenCalledWith({
      name: 'Application',
      label: 'Application',
      isSelected: true,
    });

    options.at(4).simulate('click', {});
    expect(fakeContext.updateSelection).toHaveBeenCalledWith({
      name: 'Pin',
      label: 'Pin',
      isSelected: true,
    });
  });
});
