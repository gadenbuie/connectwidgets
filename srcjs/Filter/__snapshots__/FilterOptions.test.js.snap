// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterOptions /> renders properly 1`] = `
<div
  className="rscfilter-field__options bg-white border"
>
  <div
    aria-checked={false}
    className="rscfilter-option"
    key="API"
    onClick={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    role="checkbox"
    tabIndex="0"
  >
    <div
      className="rscfilter-option__checkbox"
    />
    <span
      className="rscfilter-option__label"
    >
      API
    </span>
  </div>
  <div
    aria-checked={false}
    className="rscfilter-option"
    key="Application"
    onClick={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    role="checkbox"
    tabIndex="0"
  >
    <div
      className="rscfilter-option__checkbox"
    />
    <span
      className="rscfilter-option__label"
    >
      Application
    </span>
  </div>
  <div
    aria-checked={false}
    className="rscfilter-option"
    key="Document"
    onClick={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    role="checkbox"
    tabIndex="0"
  >
    <div
      className="rscfilter-option__checkbox"
    />
    <span
      className="rscfilter-option__label"
    >
      Document
    </span>
  </div>
  <div
    aria-checked={false}
    className="rscfilter-option"
    key="Model"
    onClick={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    role="checkbox"
    tabIndex="0"
  >
    <div
      className="rscfilter-option__checkbox"
    />
    <span
      className="rscfilter-option__label"
    >
      Model
    </span>
  </div>
  <div
    aria-checked={false}
    className="rscfilter-option"
    key="Pin"
    onClick={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    role="checkbox"
    tabIndex="0"
  >
    <div
      className="rscfilter-option__checkbox"
    />
    <span
      className="rscfilter-option__label"
    >
      Pin
    </span>
  </div>
</div>
`;
